Pivot from a resolver registry to a database registry (#6)
mnot/public-resolver-errors: draft-nottingham-public-resolver-errors.md

time and make them available.

This draft defines a mechanism to communicate an identifier in a database of
filtering incidents when a DNS resolver filtering of a name is legally mandated,
based upon the structured error data for DNS responses introduced by
{{!I-D.ietf-dnsop-structured-dns-error}}.

A consuming party (e.g., a Web browser) can use the identifier to construct a
link to the specific entry in the database provider. This enables user agents to
direct users to a location with additional context about why the filtering was
required.

Allowing DNS resolvers to inject links or user-visible messages brings unique challenges.
Because DNS resolvers are often automatically configured by unknown networks and
DNS responses are unauthenticated, these messages can come from untrusted
parties -- including attackers (e.g., the so-called "coffee shop" attack) that
leverage many users' lack of a nuanced model of the trust relationships between
all of the parties that are involved in the service they are using. Furthermore,
lowering the barrier to the presentation of messages explaining why access has
been denied by the DNS resolver risks encouraging the wider deployment of

> DNS-based censorship on the Internet.

Reviewer: It might be good to clarify that this paragraph is about what the I-D is _not_ proposing.

This draft attempts to mitigate these risks by minimising the information
carried in the DNS response to abstract, publicly registered identifiers
associated with databases of filtering incidents---the Database Operator ID and
the Filtering Incident ID. A consuming party can choose which database
identifiers they support are are willing to direct their users to, without
enabling every DNS server to surface arbitrary links and text.

## Example

In typical use, a DNS query that is filtered might contain an Extended DNS Error Code 17 (see {{!RFC8914}}) and an EXTRA-TEXT field "fdb", which is an array of references to filtering database entries:

~~~ json
{
  "fdb": [
    {"db": "example",

>     "inc": "abc123"},

Reviewer: ```suggestion
    "id": "abc123"},
```
One byte shorter?

Some organizations, such as Lumen {{lumen}}, monitor legally-mandated filtering
as a public good, and track filtering incidents in a publicly accessible
databases. Public resolvers themselves may also choose to track filtering
requests over time and make them available.

This draft defines a mechanism to communicate an identifier in a database of
filtering incidents when a DNS resolver filtering of a name is legally mandated,
based upon the structured error data for DNS responses introduced by
{{!I-D.ietf-dnsop-structured-dns-error}}.

A consuming party (e.g., a Web browser) can use the identifier to construct a
link to the specific entry in the database provider. This enables user agents to
direct users to a location with additional context about why the filtering was
required.

This abstraction is necessary because allowing DNS resolvers to inject links or

> user-visible messages brings unique challenges. DNS resolvers are often

Reviewer: ```suggestion
user-visible messages would bring unique challenges. DNS resolvers are often
```

>
> https://resolver.example.com/filtering-incidents/abc123

> Note that there is no requirement for the resolver to construct links to any database, nor for results from any DNS server. The resolver both choose which database providers it supports, and can evaluate whatever mechanisms it chooses to determine when and if to provide the link to the database provider.

Reviewer: ```suggestion
Note that there is no requirement for the resolver to construct links to any database, nor for results from any DNS server. The resolver both chooses which database providers it supports, and can evaluate whatever mechanisms it chooses to determine if and when to provide a link to the database entry.
```

This section defines the data types used to look up the details of a filtering incident from a DNS error response. Note that these identifiers are not for presentation to end users.

## DNS Filtering Database Entry {#entry-id}

A Filtering Database ID is a short, textual string that uniquely identifies the
operator of a database of filtering incidents. It uses the key "db".

A Filtering Incident ID is an opaque, string identifier for a particular
filtering incident. It might be specific to a particular request, but need not

> be. It uses the key "inc".

Author: Done.

This section defines the data types used to look up the details of a filtering incident from a DNS error response. Note that these identifiers are not for presentation to end users.

## DNS Filtering Database Entry {#entry-id}

> A Filtering Database ID is a short, textual string that uniquely identifies the

Reviewer: ```suggestion
A Filtering Database ID is a short string that uniquely identifies the
```
Or is a `string` different from a `textual string` somehow?

Author: @mnot I stole this from your original text---is there a difference?

operator of a database of filtering incidents. It uses the key "db".

> A Filtering Incident ID is an opaque, string identifier for a particular

Reviewer: ```suggestion
A Filtering Incident ID is a string identifier for a particular
```
Or define what you mean by `opaque`.

: an e-mail address or other appropriate contact mechanism

DNS Resolver Operator ID:

> : see {{entry-id}}

Reviewer 2: This section would need to be updated -- we're now identifying databases, not resolvers.

I don't think PENs would be useful here, as it would require another registry (we still need the URL templates).
